### PR TITLE
cvs-fast-export: add head, livecheck

### PR DESCRIPTION
**Pull request by jabenninghoff**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----
Add livecheck for cvs-fast-export

### CODE REVIEW

**Comment by chenrui333** [comment]
my idea would be adding a head build, update livecheck to use `:head`

**Comment by jabenninghoff** [comment]
> my idea would be adding a head build, update livecheck to use `:head`

I can do that if that is the preferred method... is it?

**Comment by chenrui333** [comment]
yeah, I will @samford weigh in before we make a change :)

File: Formula/cvs-fast-export.rb
Comment (samford) [inline Formula/cvs-fast-export.rb:9]:
```suggestion
    url :homepage
    regex(/href=.*?cvs-fast-export[._-]v?(\d+(?:\.\d+)+)\.t/i)
```

**Review by samford — CHANGES_REQUESTED** [review]
We're using an archive file that's found on the homepage as the `stable` URL, so it's more appropriate to check that page (to align the check with the `stable` source).

---

If checking the Git repository tags was appropriate here, I would have probably been fine with the Git repo URL in the `livecheck` block in this particular instance. I tried doing a build with the Git URL as `head` but it failed, so we would need to modify the formula to support building from `HEAD`. I usually try to get the `HEAD` build to work when I have time for it (I don't always succeed) but I think it would be asking too much to expect others to do the same just for the sake of a `livecheck` block.

That said, It's good to at least try this first, as sometimes building from `HEAD` works fine with no (or minimal) tweaks and having the Git repo URL as head is preferable. Other times it's more trouble than it's worth and I know there have been a handful of cases where I've used the Git URL as a string literal in a `livecheck` block.. However, this tends to be rare because it's usually not necessary (`url :stable` is often sufficient) or there are often more appropriate alternatives (as here).

**Comment by jabenninghoff** [comment]
I was able to figure out how to get a `HEAD` build working, the build requires a [newer version](https://gitlab.com/esr/cvs-fast-export/-/issues/18) of `bison`. Is this alternate update OK?

File: Formula/cvs-fast-export.rb
Comment (jabenninghoff) [reply Formula/cvs-fast-export.rb]: See my comment below - I was able to get `HEAD` working

File: Formula/cvs-fast-export.rb
Comment (carlocab) [inline Formula/cvs-fast-export.rb:23]:
```suggestion
  depends_on "bison" => :build if build.head?
```
Since you only need this for the head build?

Comment (jabenninghoff) [reply Formula/cvs-fast-export.rb:23]: Good suggestion, but reading [this issue](https://gitlab.com/esr/cvs-fast-export/-/issues/18) from `cvs-fast-export`, it's clear that the upstream expects `bison` 3 or newer; so I think it's safer (ie more future-proof) to keep this in the base build as well.

When I originally wrote the formula, I had problems with the build not working reliably without `docbook-xsl` - I now believe I missed this build dependency when I first submitted.

OK?

Comment (samford) [reply Formula/cvs-fast-export.rb:23]: `depends_on "bison" => :build if build.head?` didn't work when I tested it. `build.head?` works in the `install` block but doesn't seem to work in this outer scope, so `bison` doesn't get included when `if build.head?` is added to this statement.

The way I usually handle this is to add the `depends_on` call in a `head` block, like:

```ruby
head do
  url "https://gitlab.com/esr/cvs-fast-export.git"
  depends_on "bison" => :build
end
```

This worked as expected when I tested it.

The trouble is that the `head` block comes after the `livecheck` block in the component order, so using `url :head` in the `livecheck` block with this setup won't work. I'll be creating a Homebrew/brew PR in a moment to modify the component order, though there's no guarantee it will be merged. [If it does get merged, we'll need to update formulae with `head` blocks to move them into the new position.]

Comment (jabenninghoff) [reply Formula/cvs-fast-export.rb:23]: As I noted, `depends_on "bison" => :build` really should be part of the base formula, since the upstream package assumes `bison` version 3... is this PR acceptable as it currently stands?

Comment (samford) [reply Formula/cvs-fast-export.rb:23]: Sorry about that, I hadn't refreshed the page before commenting so I missed your previous comment.

From what I've seen, building from source using the `stable` tarball (1.55) looks exactly the same with or without `depends_on "bison" => :build`. `bison` is necessary when doing a `HEAD` build but it doesn't seem to be required when using the latest release archive.

The notable difference between the tarball from the first-party website and the files on GitLab is that GitLab is missing `gram.c`, `gram.h`, `lex.c`, and `lex.h`. Looking at the `Makefile`, Bison is involved with the `gram` files, so that makes sense.

If Bison will be required to build the next version after `1.55` (using a release tarball like we're currently doing), then it would make sense to add `depends_on "bison" => :build` without restricting it to `HEAD` builds. However, I'm not sure how we generally feel about preemptively adding dependencies before they're explicitly needed. I imagine we probably lean toward, "Wait to add the dependency until it's actually needed" (i.e., add the dependency as part of the next version bump if it's needed at that time).

That said, I'm not familiar with this formula/software, so please correct my understanding if I've misunderstood.

Comment (jabenninghoff) [reply Formula/cvs-fast-export.rb:23]: OK, given all that it does make sense to restrict the `depends_on` to just `HEAD` for now. Should I wait for https://github.com/Homebrew/brew/pull/10157 or just implement the requested changes?

Comment (samford) [reply Formula/cvs-fast-export.rb:23]: My suggested `livecheck` block above doesn't use the Git repository, so we aren't technically blocked by the brew PR. I can merge this if we do the following:

* Integrate the suggested `livecheck` block (as it aligns the check with the source of the `stable` archive and we do this whenever possible).
* Remove `head` and add the following `head` block below the `bottle` block:

```ruby
head do
  url "https://gitlab.com/esr/cvs-fast-export.git"
  depends_on "bison" => :build
end
```

File: Formula/cvs-fast-export.rb
Comment (jabenninghoff) [reply Formula/cvs-fast-export.rb]: I've updated the livecheck to use `:homepage`

File: Formula/cvs-fast-export.rb
Comment (jabenninghoff) [reply Formula/cvs-fast-export.rb]: Thanks for the clarification! I've resolved both of these, can you review?

**Review by samford — APPROVED** [review]
I tested both the `head` build and the `livecheck` block on the latest commit and everything looks good to me.

Thanks, @jabenninghoff!

**Pull request merged**
